### PR TITLE
[E2E-Improvement] Playwright page object model

### PR DESCRIPTION
**Pull request by evertonlperes**
### Changes
1. Implemented [POM](https://playwright.dev/docs/test-pom) (Page Object Model) feature.
2. Created single POM pages for each tab (isolation).
3. Refactored existing `spec` files.

### Why POM?
* It abstracts over the app pages to simplify interactions with them in multiple tests, avoiding code duplication and easy to maintain/change.
* It keeps the spec files cleaner and understandable.

### CODE REVIEW

File: e2e/main.e2e.spec.ts [hunk 1]
```diff
@@ -40,69 +43,54 @@ test.describe.serial('Main App Test', () => {
   });
 
   test('should land on General page', async() => {
-    mainTitle = page.locator(mainTitleSelector);
+    const navPage = new NavPage(page);
 
-    await expect(mainTitle).toHaveText('Welcome to Rancher Desktop');
+    await navPage.getPageTitle('Welcome to Rancher Desktop');
```
Comment (mook-as) [inline e2e/main.e2e.spec.ts:48]:
```suggestion
    await expect(navPage.pageTitle).toHaveText('Welcome to Rancher Desktop');
```

Looking at [docs](https://playwright.dev/docs/test-pom), it doesn't look like page objects are expected to embed assertions, other than as timing points…

**Review by mook-as — CHANGES_REQUESTED** [review]

File: e2e/helm.e2e.spec.ts [hunk 2]
```diff
@@ -42,13 +44,9 @@ test.describe.serial('Helm Deployment Test', () => {
...
   test('should start loading the background services', async() => {
-    const progressBarSelector = page.locator('.progress');
+    const navPage = new NavPage(page);
 
-    // Wait until progress bar show up. It takes roughly ~60s to start in CI
-    await progressBarSelector.waitFor({ state: 'visible', timeout: 200_000 });
-    // Wait until progress bar be detached. With that we can make sure the services were started
-    await progressBarSelector.waitFor({ state: 'detached', timeout: 300_000 });
-    await expect(progressBarSelector).toBeHidden();
+    await navPage.getProgressBar();
```
[the remark this answers is not in the document]
Comment (evertonlperes) [reply e2e/helm.e2e.spec.ts:49]: I'll fix that. Thanks for reviewing.

File: e2e/main.e2e.spec.ts [hunk 3]
```diff
@@ -3,8 +3,12 @@ import path from 'path';
 import {
   ElectronApplication, BrowserContext, _electron, Page, Locator
 } from 'playwright';
-import { test, expect } from '@playwright/test';
+import { test } from '@playwright/test';
 import { createDefaultSettings, playwrightReportAssets } from './utils/TestUtils';
+import { NavPage } from './pages/nav-page';
+import { K8sPage } from './pages/k8s-page';
+import { WslPage } from './pages/wsl-page';
```
[the remark this answers is not in the document]
Comment (evertonlperes) [reply e2e/main.e2e.spec.ts:10]: Yep, I thought on the same approach, but the set up take a few seconds and it will impact a bit on the test time.
Only main spec will be affected by this approach, so I'm assuming this is ok.

File: e2e/main.e2e.spec.ts (hunk 1 above)
Comment (evertonlperes) [reply e2e/main.e2e.spec.ts:48]: That interesting. It actually works, embedding asserting straight in the page model... but after you comments I realize it seems quite weird though, too much unnecessary abstraction.
thanks for the heads up, I'll change that.

File: e2e/main.e2e.spec.ts [hunk 4]
```diff
@@ -40,69 +43,55 @@ test.describe.serial('Main App Test', () => {
...
-    // Collecting data from selectors
-    const k8sSettingsTitle = page.locator(mainTitleSelector);
-    const k8sMemorySlider = page.locator(k8sMemorySliderSelector);
-    const k8sCpuSlider = page.locator(k8sCpuSliderSelector);
-    const k8sPort = page.locator(k8sPortSelector);
-    const k8sResetButton = page.locator(k8sResetBtn);
+    const navPage = new NavPage(page);
+    const k8sPage = new K8sPage(page);
+
+    await navPage.navigateTo('K8s');
```
Comment (mook-as) [inline e2e/main.e2e.spec.ts:62]: API discussion:
Would it make sense to return the navigated-to page? Something like:
```ts
const navPage = new NavPage(page);
const k8sPage = await navPage.navigateTo('K8s');
```
We'd then have to implement it as something like:
```ts
const pageConstructors = {
  'K8s': K8sPage,
  'Integrations': WslPage,
  …
};
…
    async navigateTo(tab: keyof typeof pageConstructors) {
      await this.page.click(`.nav li[item="/${ tab }"] a`);
      await this.page.waitForNavigation({ url: `**/${ tab }`, timeout: 60_000 }),
      return new (pageConstructors[tab])(this.page);
    }
```

File: e2e/main.e2e.spec.ts [hunk 5]
```diff
@@ -40,69 +43,55 @@ test.describe.serial('Main App Test', () => {
...
 
     if (!os.platform().startsWith('win')) {
-      await expect(k8sMemorySlider).toBeVisible();
-      await expect(k8sCpuSlider).toBeVisible();
+      await expect(k8sPage.k8sMemorySliderSelector).toBeVisible();
+      await expect(k8sPage.k8sCpuSliderSelector).toBeVisible();
```
Comment (mook-as) [inline e2e/main.e2e.spec.ts:66]: Hmm, shouldn't we be testing that they're _not_ visible on Windows?

File: e2e/pages/k8s-page.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Page, Locator } from 'playwright';
+import { expect } from '@playwright/test';
+
+export class K8sPage {
+    readonly page: Page;
+    readonly k8sMemorySliderSelector: Locator;
```
Comment (mook-as) [inline e2e/pages/k8s-page.ts:6]: Why are these things called `…Selector`? Compared to, say, `k8sResetBtn` (as opposed to `k8sResetBtnSelector`).

Do we even need the `k8s` prefix, given that it's hanging off the K8s page?

File: e2e/pages/k8s-page.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    readonly k8sCpuSliderSelector: Locator;
+    readonly k8sPortSelector: Locator;
+    readonly k8sResetBtn: Locator;
```
Comment (mook-as) [inline e2e/pages/k8s-page.ts:9]: This doesn't look too long; would `k8sResetButton` be okay?

**Review by mook-as — CHANGES_REQUESTED** [review]

File: e2e/main.e2e.spec.ts (hunk 4 above)
Comment (evertonlperes) [reply e2e/main.e2e.spec.ts:62]: Thanks for sharing the idea.

That's is useful to perform browser level control, as you have access to the page property. 
AFAIK playwright will keep the `page` property available after this [process](https://playwright.dev/docs/api/class-electronapplication#electron-application-first-window) : 
`page = await electronApp.firstWindow();`

`page` will be globally available across all test cases.
Returning `page` from each tab, I cannot have access to the selector from the POM page, requiring another instance t `const k8sPageItems =  new K8sPage(page);`

The way that I've implemented the `navigateTo()`, it's basically to simulate how a user will perform, navigating to the tab and using the selector `.nav`.

@mook-as 
I'm not a TS guru yet, but please, share your thoughts :)

File: e2e/main.e2e.spec.ts (hunk 5 above)
Comment (evertonlperes) [reply e2e/main.e2e.spec.ts:66]: Totally! nice catch :) 
I'll fix that

File: e2e/pages/k8s-page.ts (hunk 7 above)
Comment (evertonlperes) [reply e2e/pages/k8s-page.ts:9]: Sure thing, thanks for calling this out.

File: e2e/pages/k8s-page.ts (hunk 6 above)
Comment (evertonlperes) [reply e2e/pages/k8s-page.ts:6]: Thanks for calling this out. I'll improve it a bit more.

File: e2e/pages/portforward-page.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { Page, Locator } from 'playwright';
+import { expect } from '@playwright/test';
+
+export class PortForwardPage {
+    readonly page: Page;
+    readonly portForwardingContent: Locator;
```
Comment (mook-as) [inline e2e/pages/portforward-page.ts:6]:
```suggestion
    readonly content: Locator;
```
Since this already the Port Forwarding page.  Same for `wslDescription` in `wsl-page.ts`.

File: e2e/main.e2e.spec.ts (hunk 4 above)
Comment (mook-as) [reply e2e/main.e2e.spec.ts:62]: Hmm, let me rephrase what you said to ensure I understand you correctly:
You're saying that we would want to have access to a page without navigating to it first (because we expect it to already be there), is that correct? In that case, I'd view that as an optimization and it's better to do redundant actions in the test so that when editing the test we don't need to keep the context of related tests in our head, specially in cases like deleting a test.

> Returning `page` from each tab, I cannot have access to the selector from the POM page, requiring another instance t `const k8sPageItems = new K8sPage(page);`

Could you clarify which _selector_ you're referring to here? Given that the argument passed to `new (pageConstructors[tab])(this.page)` is just the same one that was passed to `new NavPage(page);`, I don't see how it's different from the test code calling `new K8sPage(page);` directly.  The test would then call things like `await expect(k8sPage.memorySlider).toBeVisible();` as you do now.

Thanks! Please let me know if a scratch branch to demonstrate what I'm thinking of would make things clearer.

Comment (evertonlperes) [reply e2e/main.e2e.spec.ts:62]: > Hmm, let me rephrase what you said to ensure I understand you correctly: You're saying that we would want to have access to a page without navigating to it first (because we expect it to already be there), is that correct? In that case, I'd view that as an optimization and it's better to do redundant actions in the test so that when editing the test we don't need to keep the context of related tests in our head, specially in cases like deleting a test.

Nope,
We need to navigate to the page in order to load the page elements, otherwise, we can't check the components.

> Could you clarify which selector you're referring to here? Given that the argument passed to new (pageConstructors[tab])(this.page) is just the same one that was passed to new NavPage(page);, I don't see how it's different from the test code calling new K8sPage(page); directly. The test would then call things like await expect(k8sPage.memorySlider).toBeVisible(); as you do now.

I meant the HTML components to be asserted:
```typescript
      this.mainTitle = page.locator('[data-test="mainTitle"]');
      this.progressBar = page.locator('.progress');
```

I've created this [branch](https://github.com/evertonlperes/rancher-desktop/tree/sandbox-mark-thoughts) and implemented your suggestions.

Let's discuss a bit more and really appreciate your support :)

File: e2e/helm.e2e.spec.ts [hunk 9]
```diff
@@ -42,13 +44,10 @@ test.describe.serial('Helm Deployment Test', () => {
...
-    const progressBarSelector = page.locator('.progress');
+    const navPage = new NavPage(page);
 
-    // Wait until progress bar show up. It takes roughly ~60s to start in CI
-    await progressBarSelector.waitFor({ state: 'visible', timeout: 200_000 });
-    // Wait until progress bar be detached. With that we can make sure the services were started
-    await progressBarSelector.waitFor({ state: 'detached', timeout: 300_000 });
-    await expect(progressBarSelector).toBeHidden();
+    await navPage.progressBecomesReady();
+    await expect(navPage.progressBar).toBeHidden();
```
Comment (mook-as) [inline e2e/helm.e2e.spec.ts:50]: This is duplicating the test in `main`.

**Review by mook-as — APPROVED** [review]
